chain/store/store.go

```diff
@@ -1027,44 +1087,111 @@ func (cs *ChainStore) persistBlockHeaders(ctx context.Context, b ...*types.Block
 	return err
 }
 
-func (cs *ChainStore) expandTipset(ctx context.Context, b *types.BlockHeader) (*types.TipSet, error) {
+func (cs *ChainStore) ExpandTipsetByHeightAndParent(ctx context.Context, height abi.ChainEpoch, parentTipsets []cid.Cid) (types.TipSetKey, error) {
 	// Hold lock for the whole function for now, if it becomes a problem we can
 	// fix pretty easily
 	cs.tstLk.Lock()
 	defer cs.tstLk.Unlock()
 
-	all := []*types.BlockHeader{b}
-
-	tsets, ok := cs.tipsets[b.Height]
+	blocks, ok := cs.tipsets[height]
 	if !ok {
-		return types.NewTipSet(all)
+		return types.EmptyTSK, nil
 	}
 
-	inclMiners := map[address.Address]cid.Cid{b.Miner: b.Cid()}
-	for _, bhc := range tsets {
-		if bhc == b.Cid() {
+	inclMiners := map[address.Address]*types.BlockHeader{}
+	badMiners := map[address.Address]struct{}{}
+	for _, bhc := range blocks {
+		h, err := cs.GetBlock(ctx, bhc)
+		if err != nil {
+			return types.EmptyTSK, xerrors.Errorf("failed to load block (%s) for tipset expansion: %w", bhc, err)
+		}
+
+		if _, ok := badMiners[h.Miner]; ok {
 			continue
 		}
 
+		if blkCid, found := inclMiners[h.Miner]; found {
+			log.Warnf("Have multiple blocks from miner %s at height %d in our tipset cache %s-%s, purging them all", h.Miner, h.Height, h.Cid(), blkCid)
+			badMiners[h.Miner] = struct{}{}
+			delete(inclMiners, h.Miner)
+			continue
+		}
+
+		if types.CidArrsEqual(h.Parents, parentTipsets) {
+			inclMiners[h.Miner] = h
+		}
+	}
+
+	var ret []cid.Cid
+	for _, blk := range inclMiners {
+		ret = append(ret, blk.Cid())
+	}
+
+	if len(ret) == 0 {
+		return types.EmptyTSK, nil
+	}
+
+	return types.NewTipSetKey(ret...), nil
+}
+
+func (cs *ChainStore) FormBestTipSetForHeight(ctx context.Context, height abi.ChainEpoch) (*types.TipSet, types.BigInt, error) {
```

Comment:
This should not be exported.

Reply:
Only exported for tests, which need to be in a separate package (since they depend on the chain generator, which depends on the chainstore...)

Reply:
Ugh.... ok.